src/server/game/BattlePets/BattlePetMgr.cpp

```diff
@@ -0,0 +1,419 @@
+        return;
+    }
+
+    uint32 count = 0;
+    do
+    {
+        Field* fields = result->Fetch();
+        uint32 speciesId = fields[0].GetUInt32();
+        uint16 breedId = fields[1].GetUInt16();
+
+        if (!sBattlePetSpeciesStore.LookupEntry(speciesId))
+        {
+            TC_LOG_ERROR("sql.sql", "Non-existing BattlePetSpecies.db2 entry %u was referenced in `battle_pet_breeds` by row (%u, %u).", speciesId, speciesId, breedId);
+            continue;
+        }
+
+        // TODO: verify breed id (3 - 12 (male) or 3 - 22 (male and female)) if needed
+
+        _availableBreedsPerSpecies[speciesId].insert(breedId);
+        ++count;
+    } while (result->NextRow());
+
+    TC_LOG_INFO("server.loading", ">> Loaded %u battle pet breeds.", count);
+}
+
+void BattlePetMgr::LoadDefaultPetQualities()
+{
+    QueryResult result = WorldDatabase.Query("SELECT speciesId, quality FROM battle_pet_quality");
+    if (!result)
+    {
+        TC_LOG_INFO("server.loading", ">> Loaded 0 battle pet qualities. DB table `battle_pet_quality` is empty.");
+        return;
+    }
+
+    do
+    {
+        Field* fields = result->Fetch();
+        uint32 speciesId = fields[0].GetUInt32();
+        uint8 quality = fields[1].GetUInt8();
+
+        if (!sBattlePetSpeciesStore.LookupEntry(speciesId))
+        {
+            TC_LOG_ERROR("sql.sql", "Non-existing BattlePetSpecies.db2 entry %u was referenced in `battle_pet_quality` by row (%u, %u).", speciesId, speciesId, quality);
+            continue;
+        }
+
+        // TODO: verify quality (0 - 3 for player pets or 0 - 5 for both player and tamer pets) if needed
+
+        _defaultQualityPerSpecies[speciesId] = quality;
+    } while (result->NextRow());
+
+    TC_LOG_INFO("server.loading", ">> Loaded %u battle pet qualities.", _defaultQualityPerSpecies.size());
+}
+
+uint16 BattlePetMgr::RollPetBreed(uint32 species)
+{
+    auto itr = _availableBreedsPerSpecies.find(species);
+    if (itr == _availableBreedsPerSpecies.end())
+        return 3; // default B/B
+
+    return Trinity::Containers::SelectRandomContainerElement(itr->second);
+}
+
+uint8 BattlePetMgr::GetDefaultPetQuality(uint32 species)
+{
+    auto itr = _defaultQualityPerSpecies.find(species);
+    if (itr == _defaultQualityPerSpecies.end())
+        return 0; // default poor
+
+    return itr->second;
+}
+
+BattlePetMgr::BattlePetMgr(WorldSession* owner)
+{
+    _owner = owner;
+    for (uint8 i = 0; i < MAX_PET_BATTLE_SLOTS; ++i)
+    {
+        WorldPackets::BattlePet::BattlePetSlot slot;
+        slot.Index = i;
+        _slots.push_back(slot);
+    }
+}
+
+void BattlePetMgr::LoadFromDB(PreparedQueryResult pets, PreparedQueryResult slots)
+{
+    if (pets)
+    {
+        do
+        {
+            Field* fields = pets->Fetch();
+            if (BattlePetSpeciesEntry const* speciesEntry = sBattlePetSpeciesStore.LookupEntry(fields[1].GetUInt32()))
+            {
+                uint32 species = fields[1].GetUInt32();
```

Comment:
This is getting the same field as the line above. Perhaps move this variable outside of the if?